File: blog-backend/blog-core/src/main/java/com/zach/blog/service/CategoryService.java
package com.zach.blog.service;

import com.zach.blog.dto.request.CreateCategoryRequest;
import com.zach.blog.dto.request.UpdateCategoryRequest;
import com.zach.blog.model.ApplicationUser;
import com.zach.blog.model.Category;
import org.springframework.data.domain.Page;

import java.util.List;

public interface CategoryService {
    List<Category> getCategories();

    List<Category> getAllCategories();

    Page<Category> getCategories(Integer pageNum, Integer pageSize, String name, Boolean enable);

    Category createCategory(CreateCategoryRequest request, Long userId);

    void updateCategory(Long userId, Long id, UpdateCategoryRequest request);

    void deleteCategory(Long id);

    void changeCategoryStatus(Long id, Boolean enable, Long userId);

    List<Category> getParentCategories();
}

File: blog-frontend/blog-admin/src/services/resources/category-service.ts
import { resourceAxios } from "../../utils/axios-utils"

export const categoryService = {
    getAll: async () => {
        const result = await resourceAxios.get<void, Category[]>('/categories/all')

        return result
    },

    getCategories: async ({ pageSize = 5, pageNum = 0, name, enable }: GetCategories) => {
        const result = await resourceAxios.get<void, PageRespsone<Category>>('/categories', {
            params: {
                pageSize,
                pageNum,
                ...(name !== '' ? { name } : {}),
                ...(enable !== null ? { enable } : {})
            }
        })
        return result
    },

    changeCategoryStatus: async ({ id, enable }: ChangeStatusRequest) => {
        const result = await resourceAxios.put<void, void>(`/categories/${id}/status`, {
            enable
        })

        return result
    },

    createCategory: async (data: CreateCategory) => {
        const result = await resourceAxios.post<void, Category>('/categories', data)
        return result
    },

    updateCategory: async ({ id, category }: UpdateCategory) => {
        const result = await resourceAxios.put<void, void>(`/categories/${id}`, category)

        return result
    },

    getParentCategories: async () => {
        const result = await resourceAxios.get<void, ParentCategory[]>('/categories/parent')
        return result
    },

    deleteCategory: async (id: number) => {
        const result = await resourceAxios.delete<void, void>(`/categories/${id}`)

        return result
    }

}